_drafts/elf-format-differences.md

```diff
@@ -0,0 +1,149 @@
+
+<!-- excerpt start -->
+
+The ELF object file format is one of the most commonly used today. Most build systems provide an output to this format and ELF is commonly used to output coredumps. The format varies in subtle ways for 32-bit and 64-bit targets though, which can present problems for tools supporting both. This post will highlight the main differences and is intended as a quick reference for these differences.
+
+<!-- excerpt end -->
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## ELF Background
+
+If you’re running a program on a UNIX-ish system, you’re probably using an ELF file. If you’re building firmware for your MCU, it’s probably an ELF file targeted for your MCU architecture. MacOS and Windows of course have their own slightly different formats, mach-o and Portable Executable. This format contains information on how to load your program into memory or to link with other object files to form an executable object. Static libraries are ELFs. The result of compiling a source file produces an ELF. Coredumps can be collected as ELFs. You might even be an ELF 😱!
+
+## ELF Internals
+
+This post won’t detail the complete specification, for that task I’ll recommend these:
+
+- [https://blog.k3170makan.com/2018/09/introduction-to-elf-format-elf-header.html](https://blog.k3170makan.com/2018/09/introduction-to-elf-format-elf-header.html) Excellent series with some great illustrations highlighting the specific bytes belonging to parts of the ELF
+- [https://cpu.land/becoming-an-elf-lord](https://cpu.land/becoming-an-elf-lord) A very fun overview focusing on how a program is actually run. Chapter 4 focuses on ELFs
+- [https://uclibc.org/docs/elf.pdf](https://uclibc.org/docs/elf.pdf) The original ELF-32 specification
+- [https://uclibc.org/docs/elf-64-gen.pdf](https://uclibc.org/docs/elf-64-gen.pdf) The newer ELF-64 specification
+
+## ELF-64 vs ELF-32
+
+At a high level, the two formats are structurally very similar but will differ in the following ways:
+
+- Data Representation: data types for addresses, offsets, and sizes will match their native sizes and alignments
+- Padding: Equivalent data types in ELF-64 may be packed differently compared to ELF-32 to minimize padding
+
+## Data Representation Differences
+
+The main place where the two formats differ is in how they represent data. One difficult part of looking at either standard is that data types are not defined with a naming scheme that makes the size of each type obvious, e.g. `uint32_t`. To help summarize, see these tables:
+
+| Type | Size | Alignment |
+| --- | --- | --- |
+| Elf32_Addr | 4 | 4 |
+| Elf32_Off | 4 | 4 |
+| Elf32_Half | 2 | 2 |
+| Elf32_Word | 4 | 4 |
+| Elf32_Sword | 4 | 4 |
+| unsigned char | 1 | 1 |
+| - | - | - |
+| - | - | - |
+
+| Type | Size | Alignment |
+| --- | --- | --- |
+| Elf64_Addr | 8 | 8 |
+| Elf64_Off | 8 | 8 |
```

Comment:
Maybe combine in a single table, to better highlight the differences? Or add emphasis for the ones that are different?